File: modular_exponentiation/python/mod_exp.py
# Use modular exponentiation to generate ciphertext from a given input

import sys

def mod_exp(s, r, n):
    p = 1
    while(r > 0):
        if(r % 2 == 1):
            p = p*s % n
        s = s*s % n
        r = int(r / 2)
    
    return p

s = int(input("Enter an x-value: "))
r = int(input("Enter a y-value: "))
n = int(input("Enter a mod-value: "))

print("Result:", mod_exp(s, r, n))

File: modular_exponentiation/ruby/mod_exp.rb
# Calculates Modular Exponentiation of a large integer

# Get input from user
puts "Enter an x value:"
s = Integer(gets.chomp)
puts "Enter a y value:"
r = Integer(gets.chomp)
puts "Enter a mod value:"
n = Integer(gets.chomp)
p = 1

# Calculate
while r > 0
    if r % 2 == 1
        p = p * s % n
    end
    s = s * s % n
    r = Integer(r / 2)
    #puts(r)
end

# Return the value
puts "Result: ", p